### DISCUSSION TITLE
Make it possible to evaluate arbitrary macros in the context of a given spec file

### DISCUSSION THREAD

**Discussion by michel-slm** [rpm-software-management/rpm#3008] [Ideas]
Particular use case: I'd like to be able to query a spec file for the list of its patches. Currently (on Fedora 39's 4.19.1) this is only possible using `rpmspec --shell` but it requires parsing the output:

```
❯ echo '%patches' | rpmspec --shell ./*.spec
RPM version 4.19.1.1 macro shell
> %patches
'/home/michel/src/fedora/pkgs/other/slurm/slurm_release_version.patch' '/home/michel/src/fedora/pkgs/other/slurm/slurm_check_version.patch' '/home/michel/src/fedora/pkgs/other/slurm/slurm_html_doc_path.patch' '/home/michel/src/fedora/pkgs/other/slurm/slurm_perlapi_rpaths.patch' '/home/michel/src/fedora/pkgs/other/slurm/slurm_runtime_linking.patch' '/home/michel/src/fedora/pkgs/other/slurm/slurm_to_python3.patch'
> ⏎ 
```

without shell, `rpmspec` only allows querying query tags e.g.

```
$ rpmspec --srpm --query --qf '%{name}\n' ./*.spec
slurm

$ rpmspec --srpm --query --qf '%{patches}\n' ./*.spec
error: incorrect format: unknown tag: "patches"
```

While `rpm` allows evaluating anything but only those define by shipped macros, not definitions that are spec file specific; `rpmspec` does seem to have a `-E` eval option but it is undocumented and does not seem to work

```
$ rpm -E '%_sysconfdir'
/etc

$ rpmspec -q -E '%patches' ./*.spec
error: lua script failed: attempt to index a nil value
```

**Comment by Conan-Kudo** [comment]
This would be tremendously useful, especially for me trying to dig through really complex packages... Some of them are just not able to be intuited by reading, and being able to probe them like this would be really useful!

**Comment by pmatilai** [comment]
>  I'd like to be able to query a spec file for the list of its patches.

`rpmspec --srpm -q --qf "[%{PATCH}\n]" <spec>`

**Reply by pmatilai** [reply]
The query is not exactly obvious though, so: https://github.com/rpm-software-management/rpm/pull/3011

**Comment by pmatilai** [comment]
BTW it's worth noting that both the patches and sources appear in a reverse order to how they're introduced in the spec. This is basically an internal implementation detail (linked list operation) leaking into the packages, but because it's always been that way, "fixing" would silently break anybody counting on the current behavior :-/

**Reply by Conan-Kudo** [reply]
Could you fix it for your rpmspec aliases though?

**Comment by pmatilai** [comment]
After a bit of pondering, filed https://github.com/rpm-software-management/rpm/issues/3014 instead, we'll revisit the aliases with this is fixed.
